### ISSUE TITLE
Misleading branch coverage of empty methods

### ISSUE THREAD

**Issue by nedbat** [coveragepy/coveragepy#129]
*Originally reported by* **Christian Heimes (Bitbucket: [tiran](https://bitbucket.org/tiran), GitHub: [tiran](https://github.com/tiran))**

----------------------------------------

Branch coverage of an abc.abstractmethod with just a doc string in the method, shows the decorator line as uncovered branch that never reaches "exit".

```
#!python
import abc

class AExample(object):
    @abc.abstractmethod  # exit
    def method(self):
        """doc
        """

class Example(AExample):
    def method(self):
        return True

Example().method()
```

When I replace the doc string with "pass", the decorator line is a covered branch. However the pass statement is not covered at all.

```
#!python

class AExample(object):
    @abc.abstractmethod 
    def method(self): 
        pass  # uncovered
```

----------------------------------------
- Bitbucket: https://bitbucket.org/ned/coveragepy/issue/129

**Comment by nedbat** [comment]
Turns out this isn't about abstract methods, it's about empty uncalled methods with only a docstring:

```
class Example(object):
    def method1(self):      # no branch to exit
        """doc"""

    def method2(self):
        pass                # not executed

    def method3(self):
        """doc"""
        pass                # not executed
```

**Comment by nedbat** [comment]
*Original comment by* **Alex Gaynor (Bitbucket: [alex_gaynor](https://bitbucket.org/alex_gaynor), GitHub: Unknown)**

----------------------------------------

It's unclear to me what the correct behavior is, it seems to me it's either a) no executed, or b) covered, but not a branch coverage thing. (b) is somewhat more convenient, but I'm not sure it's more correct.

**Comment by nedbat** [comment]
*Original comment by* **Mike Nerone (Bitbucket: [Manganeez](https://bitbucket.org/Manganeez), GitHub: Unknown)**

----------------------------------------

(a) would seem more appropriate for a bare function definition with only a docstring, while (b) seems right for an abstractmethod, so perhaps that should be special-cased as the OP suggested.

Incidentally, it's easy enough to do per-project because the "blamed" lines are the ones with the decorators. In your project .coveragerc:

```
#!
[report]
exclude_lines =
    @abstract
    <whatever_else>
```

**Comment by nedbat** [comment]
This will be fixed when the ast-branch code is merged.

**Comment by nedbat** [comment]
This is fixed in [44719bd (bb)](https://bitbucket.org/ned/coveragepy/commits/44719bd), which will be 4.1
